apps/settings/js/modules/addon_manager.js

```diff
@@ -0,0 +1,403 @@
+          // to the list of customizedApps
+          if (appliedCustomizations.length > 0) {
+            customizedApps.push({
+              app: app,
+              customizations: appliedCustomizations
+            });
+          }
+        });
+
+        return customizedApps;
+      });
+    },
+
+    /**
+     * Check whether an addon is enabled.
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     * @returns {Boolean}
+     */
+    isEnabled: function(addon) {
+      return this._isAddon(addon) && addon.enabled === true;
+    },
+
+    /**
+     * Check whether an addon can be deleted.
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     * @returns {Boolean}
+     */
+    canDelete: function(addon) {
+      return this._isAddon(addon) && addon.removable === true;
+    },
+
+    /**
+     * Enable an addon
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     */
+    enableAddon: function(addon) {
+      if (this._isAddon(addon)) {
+        mozApps.mgmt.setEnabled(addon, true);
+      } else {
+        return Promise.reject('not an addon');
+      }
+    },
+
+    /**
+     * Disable an addon. Returns a string indicating whether reboot or restart
+     * required for disabling the addon completely.
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     * @returns {Promise String}
+     */
+    disableAddon: function(addon) {
+      if (this._isAddon(addon)) {
+        mozApps.mgmt.setEnabled(addon, false);
+      } else {
+        return Promise.reject('not an addon');
+      }
+
+      return this._getCustomizedApps(addon).then((customizedApps) => {
+        var needsReboot = customizedApps.some((customizedApp) => {
+          var manifest = this._getManifest(customizedApp.app);
+          var role = manifest.role;
+          // If the addon affects a system app
+          if (role === 'system' || role === 'homescreen') {
+            // and the app is customized with a script, then we need reboot
+            return customizedApp.customizations.some(
+              (c) => c.scripts.length > 0);
+          } else {
+            return false;
+          }
+        });
+
+        if (needsReboot) {
+          return 'reboot';
+        }
+
+        var needsRestart = customizedApps.some((customizedApp) => {
+          return customizedApp.customizations.some((c) => c.scripts.length > 0);
+        });
+
+        if (needsRestart) {
+          return 'restart';
+        }
+
+        return 'disabled';
+      });
+    },
+
+    /**
+     * Delete an addon.
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {Object} addon
+     * @returns {Promise}
+     */
+    deleteAddon: function(addon) {
```

Comment:
If we are disabling an addon that is currently enabling, is there any problem or error that we should handle here ?